krm-functions/lib/kptrl/kptrl.go

```diff
@@ -0,0 +1,133 @@
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package kptrl
+
+import (
+	"sync"
+
+	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+)
+
+type ResourceList interface {
+	// AddResult add a result with error and corresponding kubeObject by
+	// appending a result to the result slice in the resourcelist
+	AddResult(err error, obj *fn.KubeObject)
+	// GetResults gets the results slice from the resourcelist
+	GetResults() fn.Results
+	// GetObject return an fn sdk KubeObject by comparing the APIVersion, Kind and Name
+	// if the object is found the corresponding obj is returned, if not nil is returned
+	GetObject(obj *fn.KubeObject) *fn.KubeObject
+	// GetObjects returns all items from the resourcelist
+	GetObjects() fn.KubeObjects
+	// SetObject sets the object in the resourcelist items. It either updates/overrides
+	// the entry if it exists or appends the entry if it does not exist in the resourcelist
+	// It uses APIVersion, Kind and Name to check the object uniqueness
+	SetObject(obj *fn.KubeObject)
+	// DeleteObject deletes the object from the resourcelist if it exists.
+	DeleteObject(obj *fn.KubeObject)
+}
+
+// New creates a new ResourceList interface
+// concurrency is handles in the methids
+func New(rl *fn.ResourceList) ResourceList {
+	return &resourceList{
+		rl: rl,
+	}
+}
+
+type resourceList struct {
+	m  sync.RWMutex
+	rl *fn.ResourceList
+}
+
+// AddResult add a result with error and corresponding kubeObject by
+// appending a result to the result slice in the resourcelist
+func (r *resourceList) AddResult(err error, obj *fn.KubeObject) {
+	r.m.Lock()
+	defer r.m.Unlock()
+	r.rl.Results = append(r.rl.Results, fn.ErrorConfigObjectResult(err, obj))
+}
+
+// GetResults gets the results slice from the resourcelist
+func (r *resourceList) GetResults() fn.Results {
+	r.m.RLock()
+	defer r.m.RUnlock()
+	return r.rl.Results
+}
+
+// GetObject return an fn sdk KubeObject by comparing the APIVersion, Kind and Name
+// if the object is found the corresponding obj is returned, if not nil is returned
+func (r *resourceList) GetObject(obj *fn.KubeObject) *fn.KubeObject {
+	r.m.RLock()
+	defer r.m.RUnlock()
+	for _, o := range r.rl.Items {
+		if isGVKNEqual(o, obj) {
+			return o
+		}
+	}
+	return nil
+}
+
+// GetObjects returns all items from the resourcelist
+func (r *resourceList) GetObjects() fn.KubeObjects {
+	r.m.RLock()
+	defer r.m.RUnlock()
+	return r.rl.Items
+}
+
+// SetObject sets the object in the resourcelist items. It either updates/overrides
+// the entry if it exists or appends the entry if it does not exist in the resourcelist
+// It uses APIVersion, Kind and Name to check the object uniqueness
```

Comment:
We should also use Namespace; in the case where it's local-config and has no namespace, that should still work.

Reply:
added namespace